lib/ansible/galaxy/collection/__init__.py

```diff
@@ -1597,11 +1596,17 @@ def install_artifact(b_coll_targz_path, b_collection_path, b_temp_path, signatur
 
     except Exception:
         # Ensure we don't leave the dir behind in case of a failure.
-        shutil.rmtree(b_collection_path)
+        remove_or_unlink_dir(b_collection_path)
 
         b_namespace_path = os.path.dirname(b_collection_path)
-        if not os.listdir(b_namespace_path):
-            os.rmdir(b_namespace_path)
+
+        if os.path.islink(b_namespace_path):
+            resolved_path = pathlib.Path(to_text(b_namespace_path)).resolve()
+            empty_namespace = not os.listdir(resolved_path)
+        else:
+            empty_namespace = not os.listdir(b_namespace_path)
```

Comment:
May I suggest sticking with pathlib instead of mixing low-level OS API with pathlib API? It feels cleaner this way:
```suggestion
        resolved_path = pathlib.Path(to_text(b_namespace_path))
        if resolved_path.is_symlink():
            resolved_path = resolved_path.resolve()

        empty_namespace = next(resolved_path.iterdir(), None) is None
```

(plus, it's typically recommended to operate on the pathlib objects internally and only convert to strings when paths leave the app, like when printing externally)

Reply:
By the way, can't we just resolve unconditionally here? Resolving a non-symlink path shouldn't change anything, but would hugely improve readability by removing branching…

Reply:
fixed resolved unconditionally:
```python
        resolved_namespace = pathlib.Path(to_text(b_namespace_path)).resolve()
        empty_namespace = not any(other_collection for other_collection in resolved_namespace.iterdir())
```